src/tools/paint_surface.cc

```diff
@@ -0,0 +1,259 @@
+#include <epoxy/gl.h>
+#include <glm/ext.hpp>
+
+#include "../common.h"
+#include "../ship_space.h"
+#include "../mesh.h"
+#include "../block.h"
+#include "tools.h"
+
+
+extern GLuint add_overlay_shader;
+extern GLuint remove_overlay_shader;
+extern GLuint simple_shader;
+
+extern ship_space *ship;
+
+extern hw_mesh *surfs_hw[6];
+
+bool
+paint_surface_tool::can_use(const raycast_info *rc) const {
+    auto *block = rc->block;
+    auto bl = rc->bl;
+    auto index = normal_to_surface_index(rc);
+    auto si = (surface_index)index;
+    auto *other = ship->get_block(rc->p);
+
+    // if we've started, only allow same plane
+    if (state == paint_state::started) {
+        if (other->type != block_empty) {
+            return false;
+        }
+
+        switch ((surface_index)index) {
+            case surface_xp:
+            case surface_xm:
+                if (start_block.x != bl.x && start_index != si)
+                    return false;
+                break;
+            case surface_yp:
+            case surface_ym:
+                if (start_block.y != bl.x && start_index != si)
+                    return false;
+                break;
+            case surface_zp:
+            case surface_zm:
+                if (start_block.z != bl.x && start_index != si)
+                    return false;
+                break;
+            case face_count:
+                assert(false);
+                return false;
+                break;
+        }
+    }
+
+    return (block && block->type == block_frame);
+}
+
+void
+paint_surface_tool::use(raycast_info *rc) {
+    if (!rc->hit)
+        return;
+
+    int index = normal_to_surface_index(rc);
+    auto si = (surface_index)index;
+
+    if (can_use(rc)) {
+        switch (state) {
+            case paint_state::started: {
+                finish_paint(rc->bl);
+
+                state = paint_state::idle;
+                break;
+            }
+            case paint_state::idle: {
+                start_block = rc->bl;
+                start_index = si;
+                select_type = rc->block->surfs[index];
+
+                state = paint_state::started;
+                break;
+            }
+        }
+    }
+}
+
+
+void
+paint_surface_tool::finish_paint(glm::ivec3 end) {
+    auto cur = start_block;
+
+    int i[2] = {0, 0};
+    if (start_index == surface_xm || start_index == surface_xp) {
+        i[0] = 1;
+        i[1] = 2;
+    }
+    else if (start_index == surface_ym || start_index == surface_yp) {
+        i[0] = 0;
+        i[1] = 2;
+    }
+    else {
+        i[0] = 0;
+        i[1] = 1;
+    }
+
+    auto min = glm::min(start_block, end);
+    auto max = glm::max(start_block, end);
+
+    auto s = glm::value_ptr(min);
+    auto e = glm::value_ptr(max);
+    auto c = glm::value_ptr(cur);
+    for (auto j = s[i[0]]; j <= e[i[0]]; j++) {
```

Comment:
This is a bit nasty..

Reply:
Suggestions? I don't disagree but it seemed preferable to lots of checks.